### PR TITLE
[#9626][feat] Add an auto-deploy transform for using cutlass FP4 MoE kernels

### PR DESCRIPTION
**Pull request by nzmora-nvidia**
Add a transform to replace `torch.ops.auto_deploy.torch_quant_nvfp4_moe`
with the optimized `torch.ops.auto_deploy.trtllm_quant_nvfp4_moe_fused`.

Currently generates the wrong results when the number of rows in MoE FC1 weights is not divisible by 128,
so `torch.ops.auto_deploy.trtllm_quant_nvfp4_moe_fused` is not set as the default FP4 MoE implementation (i.e. the transform is disabled).

## Release Notes

* **New Features**
  * Added support for NVFP4 Mixture of Experts quantization with FP4 weight compression for improved model efficiency.
  * Introduced configuration option to enable NVFP4 MoE fusion transformations in deployment pipeline.

* **Improvements**
  * Enhanced weight scaling and padding logic for NVFP4 quantization handling.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/_torch/auto_deploy/custom_ops/fused_moe/trtllm_moe.py [hunk 1]
```diff
@@ -234,62 +241,118 @@ def trtllm_quant_nvfp4_moe_fused(
...
+        )
+        fc2_padded[:, :, : inter_size // FP4_PER_UINT8] = fc2_expert_weights_fp4
+        fc2_expert_weights_fp4 = fc2_padded
+
+        fc2_blockscale_fp8_padded = fc2_weight_blockscale_fp8.new_zeros(
+            n_experts, hidden_size_padded, inter_size_padded // NVFP4_BLOCK_SIZE
+        )
+        fc2_blockscale_fp8_padded[:, :, : inter_size // NVFP4_BLOCK_SIZE] = (
+            fc2_weight_blockscale_fp8
+        )
```
Comment (galagam) [inline tensorrt_llm/_torch/auto_deploy/custom_ops/fused_moe/trtllm_moe.py]:
```suggestion
        assert inter_size % NVFP4_BLOCK_SIZE == 0

        fc2_blockscale_fp8_padded = fc2_weight_blockscale_fp8.new_zeros(
            n_experts, hidden_size_padded, inter_size_padded // NVFP4_BLOCK_SIZE
        )
        fc2_blockscale_fp8_padded[:, :, : inter_size // NVFP4_BLOCK_SIZE] = (
            fc2_weight_blockscale_fp8
        )
```

File: tensorrt_llm/_torch/auto_deploy/custom_ops/fused_moe/trtllm_moe.py [hunk 2]
```diff
@@ -212,17 +219,17 @@ def trtllm_quant_fp8_moe_fused_fake(
...
-    fc1_alpha: torch.Tensor,  # Precomputed FC1 alpha (1.0 / (fc1_act_global_scale * fc1_weight_blockscale_fp8))
-    fc2_alpha: torch.Tensor,  # Precomputed FC2 alpha (1.0 / (fc2_act_global_scale * fc2_weight_blockscale_fp8))
+    fc1_expert_weights_fp4: torch.Tensor,
+    fc2_expert_weights_fp4: torch.Tensor,
+    fc1_weight_blockscale_fp8: torch.Tensor,
+    fc2_weight_blockscale_fp8: torch.Tensor,
+    fc1_act_global_scale: torch.Tensor,
+    fc2_act_global_scale: torch.Tensor,
+    fc1_alpha: torch.Tensor,
+    fc2_alpha: torch.Tensor,
```
Comment (galagam) [inline tensorrt_llm/_torch/auto_deploy/custom_ops/fused_moe/trtllm_moe.py]: Why are you systematically removing comments? Here and in other files in this commit.

File: tensorrt_llm/_torch/auto_deploy/transform/library/quantization.py [hunk 3]
```diff
@@ -317,20 +320,28 @@ def quantize_weight(self, w: torch.Tensor) -> torch.Tensor:
...
         # input_scale: FP4_GLOBAL_SCALE_MAX / input_amax
         # weight_scale_2: FP4_GLOBAL_SCALE_MAX / weight_amax
         # alpha: 1 / (input_scale * weight_scale_2)
         return {
             "input_scale": torch.tensor(1.0 / 6.0),
-            "weight_scale": torch.empty((padded_m * padded_n), dtype=torch.uint8),
+            "weight_scale": torch.empty((padded_m, padded_n), dtype=torch.uint8),
+            # "weight_scale": torch.empty((m, n), dtype=torch.uint8),
+            # "weight_scale": torch.empty(padded_m * padded_n, dtype=torch.float8_e4m3fn),
+            # "weight_scale": torch.empty(padded_m * padded_n, dtype=torch.uint8),
```
Comment (galagam) [inline tensorrt_llm/_torch/auto_deploy/transform/library/quantization.py]:
```suggestion
```

**Review by galagam — APPROVED** [review]
Some general comments:

1. In a previous PR you changed the terminology from w1, w2, w3 to fc1 expect weights and fc2 expert weights, but here we're still using the old terminology. I find it confusing. I warrants at least a comment.
2. PR description states "Currently generates the wrong results when the number of rows in MoE FC1 weights is not divisible by 128"
Better check for this condition and assert. Easier to figure out what is happening when an assertion fails instead of debugging accuracy issues.
Even better if we also have a unit test for scenario, skipped and referencing a bug ID.

File: tensorrt_llm/_torch/auto_deploy/transform/library/fused_moe.py [hunk 4]
```diff
@@ -1572,3 +1572,229 @@ def _apply(
...
+        #         )
+        #     )
+        #     .to(torch.float8_e4m3fn)
+        #     .contiguous()
+        # )
+
+        ###
+        w1_alpha_stacked = _stack(w1_alpha, dim=0)
+        w2_alpha_stacked = _stack(w2_alpha, dim=0)
+        # w3_alpha_stacked = _stack(w3_alpha, dim=0)
```
[the remark this answers is not in the document]
Comment (nzmora-nvidia) [reply tensorrt_llm/_torch/auto_deploy/transform/library/fused_moe.py]: not ready yet.
The case using w3 is not in the critical path for nano so I waited with it

**Comment by nzmora-nvidia** [comment]
> Some general comments:
> 
> 1. In a previous PR you changed the terminology from w1, w2, w3 to fc1 expect weights and fc2 expert weights, but here we're still using the old terminology. I find it confusing. I warrants at least a comment.
> 2. PR description states "Currently generates the wrong results when the number of rows in MoE FC1 weights is not divisible by 128"
>    Better check for this condition and assert. Easier to figure out what is happening when an assertion fails instead of debugging accuracy issues.
>    Even better if we also have a unit test for scenario, skipped and referencing a bug ID.

1. I think you misunderstood the code - Let me know which code you're looking at. The operator is `trtllm_quant_nvfp4_moe_fused` and it's using fc1, fc2 (I created this interface in a previous PR and this PR changes it). 
2. Yeah, maybe the assert is a good idea. There is a UT with the scenario and it is skipped (`tests/unittest/_torch/auto_deploy/unit/singlegpu/custom_ops/test_trtllm_moe.py`). There's no bug id because I don't consider the task complete without this configuration working. Nonetheless, if I merge this PR in its current state then I will open a bug.

File: tensorrt_llm/_torch/auto_deploy/custom_ops/fused_moe/trtllm_moe.py (hunk 1 above)
Comment (nzmora-nvidia) [reply tensorrt_llm/_torch/auto_deploy/custom_ops/fused_moe/trtllm_moe.py]: added - thanks

File: tensorrt_llm/_torch/auto_deploy/transform/library/fused_moe.py [hunk 5]
```diff
@@ -1572,3 +1572,229 @@ def _apply(
...
+            else torch.empty(
+                0, device=w1_input_scale_stacked.device, dtype=w1_input_scale_stacked.dtype
+            )
+        )
+        # assert torch.all(w1_input_scale_stacked[0] == w1_input_scale_stacked), (
+        #     "All w1 scales should have the same value."
+        # )
+        # assert torch.all(w2_input_scale_stacked[0] == w2_input_scale_stacked), (
+        #     "All w2 scales should have the same value."
+        # )
```
[the remark this answers is not in the document]
Comment (nzmora-nvidia) [reply tensorrt_llm/_torch/auto_deploy/transform/library/fused_moe.py]: fixed - thanks

**Comment by nzmora-nvidia** [comment]
> > Some general comments:
> > 
> > 1. In a previous PR you changed the terminology from w1, w2, w3 to fc1 expect weights and fc2 expert weights, but here we're still using the old terminology. I find it confusing. I warrants at least a comment.
> > 2. PR description states "Currently generates the wrong results when the number of rows in MoE FC1 weights is not divisible by 128"
> >    Better check for this condition and assert. Easier to figure out what is happening when an assertion fails instead of debugging accuracy issues.
> >    Even better if we also have a unit test for scenario, skipped and referencing a bug ID.
> 
> 1. I think you misunderstood the code - Let me know which code you're looking at. The operator is `trtllm_quant_nvfp4_moe_fused` and it's using fc1, fc2 (I created this interface in a previous PR and this PR changes it).
> 2. Yeah, maybe the assert is a good idea. There is a UT with the scenario and it is skipped (`tests/unittest/_torch/auto_deploy/unit/singlegpu/custom_ops/test_trtllm_moe.py`). There's no bug id because I don't consider the task complete without this configuration working. Nonetheless, if I merge this PR in its current state then I will open a bug.

For (2): opened a bug and added the remarks and the assert - thanks!

File: tensorrt_llm/_torch/auto_deploy/transform/library/fused_moe.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (nzmora-nvidia) [reply tensorrt_llm/_torch/auto_deploy/transform/library/fused_moe.py]: fixed - thanks
